runtime/src/main/java/org/corfudb/runtime/collections/streaming/DeltaStream.java

```diff
@@ -0,0 +1,153 @@
+            trimMark = streamAddressSpace.getTrimMark();
+
+            // if the task is syncing it will either discover the trim during reading the addresses that
+            // are already buffered, or discover this trim the next time the task gets scheduled
+
+            // gap in stream
+            // error out
+
+            // since refresh can be called while the task is syncing we can error out when the scheduling tries to
+            // reschedule this task
+            return;
+        }
+
+        // if lastread == -1 and trim mark is larger than the address?
+        Preconditions.checkState(trimMark <= streamAddressSpace.getTrimMark());
+        trimMark = streamAddressSpace.getTrimMark();
+
+        int maxAddressesToTransfer = availableSpace();
+        long[] newAddresses = streamAddressSpace.toArray();
+        log.info("maxTotxn {} newaddress {} id {}", maxAddressesToTransfer, newAddresses, streamId);
+        for (int idx = 0; idx < newAddresses.length && idx < maxAddressesToTransfer; idx++) {
+            // if maxAddressesToTransfer doesn't fit, then the rest of the addresses will be
+            // thrown away and re-discovered on the next pool
+            Preconditions.checkState(maxAddressSeen < newAddresses[idx],
+                    "maxAddressSeen %s not < %s", maxAddressSeen, newAddresses[idx]);
+            Preconditions.checkState(addressesToRead.add(newAddresses[idx]),
+                    "failed to queue %d", newAddresses[idx]);
+            // Only update max seen after it has been added to the buffer
+            maxAddressSeen = newAddresses[idx];
+            log.info("transfering {} id {}", newAddresses, streamId);
+        }
+    }
+
+
+    // called by the syncing thread
+    // hasNext signals that there is an event to be produced, this even is either a logdata read or an exception that
+    // needs to be propagated
+    public boolean hasNext() {
+        return !addressesToRead.isEmpty() || lastAddressRead < trimMark && lastAddressRead != Address.NON_ADDRESS;
+    }
+
+    // called by the syncing thread
+    public ILogData next() {
+
+        if (lastAddressRead != Address.NON_ADDRESS && lastAddressRead < trimMark) {
+            // log trace state of buffer?
+            // test this
+            throw new TrimmedException(String.format("lastAddressRead %d trimMark %d", lastAddressRead, trimMark));
+        }
+
+        long readAddress = (long) addressesToRead.remove();
+        Preconditions.checkState(lastAddressRead < readAddress,
+                "%d must be greater than %d", readAddress, lastAddressRead);
+
+        if (readAddress <= trimMark) {
+            throw new TrimmedException(String.format("current %d trimMark %d", readAddress, trimMark));
+        }
+
+        // set cache options here
+        // test trimmed exception
+        // test hole
+        // test empty?
+        ILogData logData = addressSpaceView.read(readAddress);
```

Comment:
This read will use the default cache setting and pollute the cache for the whole runtime. Need to pass read options.

Reply:
AddressSpaceView::read will peek-sleep-read-hole fill on a shared worker thread which will block other threads that can retrieve their deltas on the first peek. A new API should be introduced to break these protocol read operations into non blocking operations so that they can be scheduled explicitly.